### ISSUE TITLE
Remove special cased OUT_DIR for `bench`, `test`, and `dev` profiles

### ISSUE THREAD

**Issue by banool** [rust-lang/cargo#11053]
### Problem

Today, `OUT_DIR` for custom user profiles and the pre-included `release` profile uses the name of the profile. For example, for a custom `performance` profile, the path might look like this:
```
/code/core/target/performance/build/my-build-info-9f91ba6f99d7a061/out
```

But for the included `bench` profile, the path looks like this:
```
/code/core/target/performance/release/my-build-info-9f91ba6f99d7a061/out
```

You'd expect it to contain `bench` instead of `release`.

In [this page](https://doc.rust-lang.org/cargo/guide/build-cache.html) in the Cargo docs, it says this is for historical reasons:
> For historical reasons, the dev and test profiles are stored in the debug directory, and the release and bench profiles are stored in the release directory. User-defined profiles are stored in a directory with the same name as the profile.

### Proposed Solution

The proposed solution is to remove the special casing and just use the profile name in all cases. This would make it easier to reliably get the profile name as part of the build, which is currently only possible by looking at `OUT_DIR` (e.g. like in [this SO question](https://stackoverflow.com/questions/73595435/how-to-get-profile-from-cargo-toml-in-build-rs-or-at-runtime/73603419#73603419)).

**Comment by weihanglo** [comment]
Yep there are some historical reasons but also benefits, as builds between dev and test profiles can share build caches and artifacts.

If we do want to go on for this, to maintain backward-compatible, maybe we can crate a symlink from new profile directory to the old one. Thought I wonder if #11054 eventually got implemented, do you still need this feature?

**Comment by jschwe** [comment]
> Thought I wonder if https://github.com/rust-lang/cargo/issues/11054 eventually got implemented, do you still need this feature?

[corrosion](https://github.com/corrosion-rs/corrosion) is a tool which integrates cargo into CMake, which includes copying build artifacts from the output directory to the location CMake expects them. #11054 would not help us in this regard, since CMake does not benefit from the knowledge a build script may have.

**Comment by epage** [comment]
Sounds like the use case would be best served by #6790.  Considering the compatibility concerns, there doesn't seem a path forward for this.  I'm going to close in favor of #6790.  If there is a reason we should re-evaluate, let us know!.
